domain/application/state/storage.go

```diff
@@ -118,6 +135,226 @@ VALUES ($storageToAdd.*)`, storageToAdd{})
+			// a pool name or a valid storage type for the relevant cloud.
+			if uuid, ok := poolsByName[t.params.PoolNameOrType]; ok {
+				inst.StoragePoolUUID = &uuid
+			} else {
+				inst.StorageType = &t.params.PoolNameOrType
+			}
+
+			if err := st.createUnitStorageInstance(ctx, tx, unitUUID, inst); err != nil {
+				return errors.Capture(err)
+			}
+
+			// TODO(storage) - insert data for the unit's assigned machine when that is implemented
+
+			// For CAAS models, we create the storage with the unit
+			// as there's no machine for the unit to be assigned to.
+			if modelType != model.CAAS {
+				continue
+			}
+
+			err = st.attachStorageToUnit(ctx, tx, inst.StorageUUID, unitUUID)
+			if err != nil {
+				return errors.Errorf("attaching storage %q to unit %q: %w", inst.StorageUUID, unitUUID, err)
+			}
+
+			// Get the info needed to create the necessary filesystem and/or volume attachments to the net node.
+			// The required attachments then inform the creation of the filesystem and/or volume.
+			filesystem, volume, err := st.attachmentParamsForNewStorageInstance(storageParentDir, inst.StorageID, t.params.PoolNameOrType, inst.StorageName, t.meta, poolKinds)
+			if err != nil {
+				return errors.Errorf("creating storage parameters: %w", err)
+			}
+			if filesystem != nil {
+				filesystemUUID, err := st.createFilesystem(ctx, tx, instUUID, netNodeUUID)
+				if err != nil {
+					return errors.Errorf("creating filesystem for storage %q for unit %q: %w", instUUID, unitUUID, err)
+				}
+				filesystem.filesystemUUID = filesystemUUID
+				if err := st.attachFilesystemToNode(ctx, tx, netNodeUUID, *filesystem); err != nil {
+					return errors.Errorf("attaching filesystem to storage %q for unit %q: %w", instUUID, unitUUID, err)
+				}
+			}
+			if volume != nil {
+				volumeUUID, err := st.createVolume(ctx, tx, instUUID, netNodeUUID)
+				if err != nil {
+					return errors.Errorf("creating volume for storage %q for unit %q: %w", instUUID, unitUUID, err)
+				}
+				volume.volumeUUID = volumeUUID
+				if err := st.attachVolumeToNode(ctx, tx, netNodeUUID, *volume); err != nil {
+					return errors.Errorf("attaching volume to storage %q for unit %q: %w", instUUID, unitUUID, err)
+				}
+			}
+		}
+	}
+	return nil
+}
+
+func (st *State) createUnitStorageInstance(ctx context.Context, tx *sqlair.TX, unitUUID coreunit.UUID, inst storageInstance) error {
+	insertStorageStmt, err := st.Prepare(`
+INSERT INTO storage_instance (*) VALUES ($storageInstance.*)
+`, storageInstance{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+
+	insertStorageUnitStmt, err := st.Prepare(`
+INSERT INTO storage_unit_owner (*) VALUES ($storageUnit.*)
+	`, storageUnit{})
+	if err != nil {
+		return errors.Capture(err)
+	}
+	err = tx.Query(ctx, insertStorageStmt, inst).Run()
+	if err != nil {
+		return errors.Errorf("creating storage instance %q for unit %q: %w", inst.StorageUUID, unitUUID, err)
+	}
+
+	storageUnit := storageUnit{
```

Comment:
Can this be declared above the statement and used in the prepare.

Reply:
ok